llama.cpp

```diff
@@ -5726,6 +5734,340 @@ static void llm_build_kv_store(
     ggml_build_forward_expand(graph, ggml_cpy(ctx, v_cur_t, v_cache_view));
 }
 
+static struct ggml_tensor * llama_build_mat_mul_blocked_computation(
```

Comment:
Is this still necessary after the fixes to the CUDA backend?

Reply:
For CUDA, I think no (I haven't verified the CUDA work done by others here yet). On MPS, you get some kind of corruption without it, nature of which I haven't fully determined (but smells a lot like another `int` overflow problem...somewhere.). Also, the blocked computation is also clearly corrupting something, but in a way that doesn't seem apparent in generating text. (`perplexity` tool goes mad). Quality is a lot better on MPS with it than without it.

I would not put that function in a PR at this time until I understand what's going wrong with it. I wrote it with the thinking "it's too much effort to chase down every GPU backend to hunt any overflow bugs; maybe it's easier to just work around it at this time" but that might have been a misjudgement.

IMO if CPU and Cuda work and there's no known other easy hanging fruit without this blocked multiplication crap then I'd approve a PR like that. I would keep chasing the MPS problems on the side and maybe fix them separately. We can track overflow issues separately.

@Carolinabanana I would consider removing the blocked multiplication matrix and related code and just keep the plain `ggml_mul_mat` in the `build_command_r()` at this time...unless by some miracle I happen to figure out what is it doing wrong on a short notice (not likely to happen because I won't have time to work on this much this weekend, but possible).

My branch is a bit experimental and I could have communicated better what is experimental parts and what are the essential parts.

Reply:
By MPS do you mean the Metal backend? I noticed that are some `uint` in `kernel_mul_mm_impl` that could be causing the issue.

Reply:
Ah right right. Yes, I mean Metal (sorry I'm used to Torch terminology). Naive search & replace `uint` to `uint64_t` upgrade didn't fix it in the `mm` code, but I spent like 1% of time trying to debug the metal code and 99% writing of my time with this crap blocked multiplication thing that doesn't work correctly :D

Reply:
Also, the code in the PR is an older version of the blocked multiplication code. I have a newer one that avoids making placeholder tensors, and does concatting a bit more efficiently but that does not fix the corruption issues unfortunately.

I noticed that depending on how you concatenate tensors, it'll give different results (the results should be 100% identical). I also thought the concatenation as written is probably busted; both in this code and in my newer code in my branch which I suspect is the part that's currently broken and causing wrong results. But can't confirm until I have time to dive into it. It's supposed to stitch a 2D grid of tensors into one big tensor but it's actually just concatenating in 1D with hacky flattening views and I think it's doing it wrong.

Reply:
@slaren @Noeda Agree, I've now removed this as it still has issues, would be complex to debug, and could affect other Llama arches as it's in the common code.